projects/views.py

```diff
@@ -58,6 +47,52 @@ class UserNotPIEligible(Exception):
     pass
 
 
+class UserPermissions():
+    manage: bool
+    manage_membership: bool
+
+    def __init__(self, manage, manage_membership):
+        self.manage = manage
+        self.manage_membership = manage_membership
+
+    @staticmethod
+    def _get_user_project_role_scopes(keycloak_client, username, project):
+        role, scopes = keycloak_client.get_user_project_role_scopes(
+            username, get_charge_code(project)
+        )
+        logger.info("SCOPES===")
+        logger.info(role)
+        logger.info(scopes)
+        return role, scopes
+
+    @staticmethod
+    def _parse_scopes(scopes):
+        # NOTE We have these two scopes defined in KC. In implementation,
+        # we do not differentiate between them. See our docs for the specifics
+        # https://chameleoncloud.readthedocs.io/en/latest/user/project.html#user-roles
+        can_manage = "manage" in scopes or "manage-membership" in scopes
+        return UserPermissions(
+            manage=can_manage,  # Can edit project metadata/allocations
+            manage_membership=can_manage,  # Can update members
+        )
+
+    @staticmethod
+    def get_user_permissions(keycloak_client, username, project):
+        _, scopes = UserPermissions._get_user_project_role_scopes(
+            keycloak_client, username, project)
+        return UserPermissions._parse_scopes(scopes)
+
+    @staticmethod
+    def get_manager_projects(keycloak_client, username):
+        """Gets project names for all project this user is a manager of
+        """
+        return [
+            project["groupName"]
+            for project in keycloak_client.get_user_roles(username)
+            if UserPermissions._parse_scopes(project["scopes"]).manage
+        ]
+
+
```

Comment:
this looks like it's centralizing the mapping from keycloak scopes -> portal permissions, rather than a bunch of if-then? scattered around?

Reply:
Yes. Now all of the permissions/scope code is here, and a block of code can check against this class for a portal permission. This should prevent bugs with manually parsing/interacting with keycloak scopes each time we check a user's role.